api/processors/actions/group/finish.py

```diff
@@ -0,0 +1,76 @@
+from django.db import transaction
+
+from api.models.auxil.status_setter import StatusSetter
+from api.models.choices.log_event_type import (
+    CoordinatorLogEventType,
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.status import (
+    GroupProjectStatus,
+    StudentProjectStatus,
+    TeacherProjectStatus,
+)
+from api.processors.actions.group import GroupActionProcessor
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupFinishProcessor(GroupActionProcessor):
+    @transaction.atomic
+    def process(self) -> None:
+        self._set_statuses()
+        self._create_log_events()
+        self._move_related_people_to_former()
+
+    def _create_log_events(self) -> None:
+        GroupLogEventCreator.create(
+            group=self.group,
+            student_log_event_type=StudentLogEventType.GROUP_FINISHED,
+            teacher_log_event_type=TeacherLogEventType.GROUP_FINISHED,
+            coordinator_log_event_type=CoordinatorLogEventType.GROUP_FINISHED,
+            group_log_event_type=GroupLogEventType.FINISHED,
+            from_group=self.group,
+        )
+
+    def _move_related_people_to_former(self) -> None:
+        teachers_current, students_current, coordinators_current = (
+            self.group.teachers,
+            self.group.students,
+            self.group.coordinators,
+        )
+        self.group.teachers_former.add(*teachers_current.all())
+        self.group.students_former.add(*students_current.all())
+        self.group.coordinators_former.add(*coordinators_current.all())
+        self.group.teachers.clear()
+        self.group.students.clear()
+        self.group.coordinators.clear()
+        self.group.save()
+
+    def _set_coordinators_status(self) -> None:
+        StatusSetter.update_statuses_of_active_coordinators(self.timestamp)
+
+    def _set_group_status(self) -> None:
+        StatusSetter.set_status(
+            obj=self.group, project_status=GroupProjectStatus.FINISHED, status_since=self.timestamp
+        )
+
+    def _set_teachers_status(self) -> None:
+        self.group.teachers_with_no_other_active_groups().update(
+            project_status=TeacherProjectStatus.WORKING,
+            situational_status="",
+            status_since=self.timestamp,
+        )
+
+        self.group.teachers_with_other_active_groups().update(
+            project_status=TeacherProjectStatus.NO_GROUP_YET,
+            situational_status="",
+            status_since=self.timestamp,
+        )
+
+    def _set_students_status(self) -> None:
+        self.group.students.update(
+            project_status=StudentProjectStatus.STUDYING,
```

Comment:
Btw, @lemontree210  are you sure it should be `studying`?

Reply:
Thanks for the attentiveness! I've been thinking about this. Theoretically it is possible that the student is studying in another group. OK this can be checked, but that is not all. If not, we don't know (at this point, when the group is just marked as finished) whether the student is going to leave the project or continue in a more advanced group. This will have to be clarified by the coordinator. So we can either set `STUDYING` or maybe we could add a `project_status` called something like `AWAITING_NEXT_STEP`, which would clearly mark the student as being in a sort of "limbo" in terms of whether they're leaving or should go back to `AWAITING_OFFER` status again, maybe after the coordinator adjusts their language level.

Reply:
Get it, then we should keep it studying and mark `finished` once the student leaves permanently

Reply:
On the other hand, you can filter by `AWAITING_NEXT_STEP` and/or put these students somewhere prominently on the dashboard. Of course if we set `STUDYING`, we can display on the dashboard the students that have project_status `STUDYING` but no groups. My point is, however, that no matter which way we choose, we will still need to keep track of these students.

Reply:
Let's ask the dashboard man @dont-text-me :) What do you think?

Reply:
good question! i think we can make do without a limbo status and just go off the combination of the most recent log event and the project status to filter out students that have just "graduated". In terms of presenting it, the simplest and most MVP friendly approach would be to create a new tab in the dashboard (see super high quality mockup attached 😄 ) that would get the list of students already filtered from the backend, since log events are not sent to the dashboard yet. In the future though, i definitely see us wanting a separate page for all the things needing attention displayed as separate widgets, but that's a bit beyond the MVP
<img width="1502" alt="image" src="https://github.com/samanthasgroup/django-webapps/assets/79531889/a7239fa0-c17a-4bfb-a339-bd4c0ccf535c">

Reply:
so overall, the simplest action we could take from this is to add a query parameter to the `/dashboard/students` view (smth like `groupRecentlyFinished`) that if set, will return filter out students that have `GROUP_FINISHED` as their latest log event + a `STUDYING` project status. I'll then create a tab in the students table to display them all

Reply:
Should we do it in a separate PR or here @lemontree210?

Reply:
> Should we do it in a separate PR or here @lemontree210?

No limbo status then and I'd say a separate PR (and I'll think about the task a bit more). Thanks @dont-text-me :)

Reply:
Created issue #237